mlir/docs/MemrefMergeAlloc.md

```diff
@@ -0,0 +1,464 @@
+calculated in the `MemorySchedule`. This step is abstracted in a
+`MemoryMergeMutatorFunc` function.
+
+```c++
+using MemoryMergeMutatorFunc = std::function<LogicalResult(
+    Operation *toplevel, Block *scope, const MemorySchedule &,
+    const MergeAllocationOptions &)>;
+```
+
+
+## Tick-based Implementation for merge-alloc
+
+A tick-based implementation of merge-alloc in provided by default. The basic
+idea of the tick-based allocation merging is that
+
+1. Each of the operations in a function is assigned a "tick". An operation with
+   a smaller tick is expected to be executed before one with a larger tick
+2. Collect the first referenced tick and the last referenced tick for each
+   mergeable allocation. If a buffer is referenced in loops and branches,
+   special handling is needed.
+3. For each allocation scope, linearize the first referenced tick and the last
+   referenced tick of mergeable allocations inside of it into a single linear
+   timeline
+4. Use a "static-memory-planner" to handle the linear timeline
+
+Limitations of Tick-based merge-alloc:
+ * only contiguous, static shaped and identical layout memrefs are considered.
+   Others are disregarded
+ * only `RegionBranchOpInterface` operations are
+   allowed to access memref inside the operations' children regions. Other
+   operaions containing regions should not access memref inside. Otherwise, a
+   pass error could occur.
+
+### Basic concepts
+
+In the context of tick-based merge-alloc, mergeable allocation and allocation
+scope are defined as follows
+
+#### Mergeable allocation
+
+The pass should only consider to merge a `memref.alloc` only if
+ * the ownership of the memref does not escape from the function or the body of
+   the loop. That is, the memref and its alias should not be returned or
+   yielded by a function or a loop.
+ * and the memref is "dense" in its strides (points to a contiguous range of
+   memory) and it has static shape
+
+In tick-based merge-alloc, we call these `memref.alloc` **mergeable**
+allocations.
+
+The memrefs passed by function arguments, or returned by the function will be
+untouched by this optimization.
+
+#### Allocation scopes
+
+The transformation first needs to identify the allocation scopes, which are
+single basic blocks of parent operaions which
+ * implement `AutomaticAllocationScope`
+ * and are not `scf.for` (allocations in an `scf.for` can be hoisted to
+ parent `AutomaticAllocationScope`)
+
+For example, below is an example IR of a function with nested `scf.forall` ops.
+
+```mlir
+func.func @mlp(...) { // <---- alloc scope 1
+   scf.for(...) { // <---- NOT an alloc scope!
+      // allocation inside will be merge to alloc scope 1 above
+   }
+   ...
+   scf.forall(...) { // <---- alloc scope 2
+      ...
+      // allocation here will be merge to alloc scope 2
+      %buf = memref.alloc() : ...
+      scf.forall(...) { // <---- alloc scope 3
+      }
+   }
+}
```

Comment:
Please re-think the behavior of `memref.alloc` in a loop. Please note that we are handling the IR before "memref-deallcation" pass, so there will be no explicit `dealloc` in the input IR of our pass.

Consider the IR before deallocation pass:

```mlir
scf.for(...) {
  %0 = memref.alloc(...)
  use(%0)
  // no use of %0 afterwards
}
```

After ` --buffer-deallocation-pipeline`, it finally outputs:

```mlir
scf.for(...) {
  %0 = memref.alloc(...)
  use(%0)
  memref.dealloc %0
}
```

So the actual meaning of `%0 = memref.alloc(...)` in an `scf.for` is to allocate a memref within the body of the for loop, and deallocate at the end of the body, **before the next iteration**.

In this case, `%0` of different iterations do **not** have overlapped lifetime. We can always safely hoist the allocation out of the loop. That is what `buffer-loop-hoisting` pass does, and the hoisting is also conducted by our pass too.

Please re-consider the original semantic of the original IR. It does allocate a new buffer at each iteration, but as long as the buffer's ownership does not escape from the loop, the buffer will be deallocated in the same iteration.

However, there will be some special case pointed out by @matthias-springer, when the memref is the loop-carried variable of the loop. In that case, the buffer of the last iteration will be used in the next iteration, and even in the result values of the `scf.for`. If that case is detected, we simply leave the allocation unchanged by this pass.

Reply:
Thanks for your explanation. This situation can also occur in other control flow scenarios. Any alias value whose lifetime can reach alloc will cause problems in essence. Does our patch need to find all possible scenarios? I think this may introduce a new analysis instead of a simple regionbranchinterface that can be handled.

Reply:
> This situation can also occur in other control flow scenarios.

Could you please provide an example of such cases? Since our pass only handles `RegionBranchOpInterface`, I think we are only handing the simple cases of structured-control-flow-like ops.

> Any alias value whose lifetime can reach alloc will cause problems in essence

This is already handled by `BufferViewFlowAnalysis`, which is used in our pass. A memory access on a memref contributes to all allocations it may aliases with.

Reply:
> Could you please provide an example of such cases?

Sorry for not make myself clear, consider the following input:
```
   ...
   ^bb0(%arg0: memref<16xf32>): 
      %alloc = memref.alloc() : memref<16xf32>
      "foo.write"(%alloc)
      "foo.read"(%arg0)
      cf.br ^bb0(%alloc : memref<16xf32>)
   ...
```

The alias of `%alloc` : `%arg0`  has a lifetime that  canreach the definition  of `%alloc`, which will also cause error as loop-carried variable.  If we have an alloc value A, and any alias value of A  has a lifetime that can reach the definition of A will cause same error, not only the branch case above or the loop case that matthias mentioned.

Reply:
We only support `RegionBranchOpInterface` in tick-based, and `cf.br` is unstructured control flow and it is unsupported.